feat(Widgets): smooth drag behavior

floryst opened the pull request:
<!--
👋 Hello, and thank you for starting this contribution!
📖 Make sure you've read our [CONTRIBUTING.md](https://github.com/Kitware/vtk-js/blob/master/CONTRIBUTING.md) guide before submitting your pull request.
❗️ Please follow the template below to help other contributors review your work.
-->

### Context
Dragging widget handles has a jerky behavior, as the handles jump to the mouse position.

### Results
<!--
Describe or illustrate the effects of your contribution. Please include:
- comparisons of the behavior before vs after
- screenshots of new or changed visualizations if applicable
-->
The manipulators now return a new property, `worldDelta`, which is the 3D vector delta between the current and the previous call to `handleEvent`.

### Changes
<!--
Please describe what is changing. Include:
- APIs added, deleted, deprecated, or changed
- Classes and methods added, deleted, deprecated, or changed
- A summary of usage if this is a new feature or change to an API. Adequate documentation and TS definitions should also be added/updated.
-->
- [x] Documentation and TypeScript definitions were updated to match those changes
- [x] Widgets have been updated to use `handleEvent().worldDelta` for smoother dragging

### PR and Code Checklist
<!--
NOTE: We will not merge if the following steps have not been completed!
-->
- [x] [semantic-release](https://github.com/semantic-release/semantic-release) commit messages
- [x] Run `npm run reformat` to have correctly formatted code

### Testing
<!--
Please describe how this can be tested by reviewers. Be specific about anything not tested and the reasons why.
Tests should complete without errors. See CONTRIBUTING.md
-->
- [ ] This change adds or fixes unit tests <!-- Tests should be added for new functionality -->
- [ ] Tested environment:
  - **vtk.js**: <!-- ex: 14.0.0 (favor latest master) -->
  - **OS**: <!-- ex: Windows 10, iOS 13.6 -->
  - **Browser**: <!-- ex: Chrome 89.0.4389.128 -->

<!--
Edit and uncomment the section below if relevant

### Funding
This contribution is funded by [Example](https://example.com).

 -->

finetjul commented on Sources/Widgets/Core/DragHelper/index.js:
```diff
@@ -0,0 +1,44 @@
+    return subtract(dragWorldCoords, model.lastOrigin, [0, 0, 0]);
+  };
+
+  publicAPI.calcPosition = (position, dragWorldCoords) => {
```
you could provide an optional 3rd element that will contain the result array. (for performance reasons)

finetjul commented on Sources/Widgets/Core/DragHelper/index.d.ts:
```diff
@@ -0,0 +1,61 @@
+  calcDelta(dragWorldCoords: Nullable<number[]>): Vector3;
+
+  /**
+   * Calculates the new position from a starting position given a drag interaction coordinate.
```
Can you give more explanations in this documentation ? It's not really clear what it does.

finetjul commented on Sources/Widgets/Core/DragHelper/index.d.ts:
```diff
@@ -0,0 +1,61 @@
+
+  /**
+   * Calculates the new position from a starting position given a drag interaction coordinate.
+   * @param {Nullable<Vector3>} startPosition the start position
```
Please details what "startPosition" is:
 - last known origin of the widget?
 - first origin of the widget when mouse was first clicked
 - ...

finetjul commented on Sources/Widgets/Core/DragHelper/index.d.ts:
```diff
@@ -0,0 +1,61 @@
+}
+
+/**
+ * Method used to decorate a given object (publicAPI+model) with vtkPlane characteristics.
```
incorrect doc

finetjul commented on Sources/Widgets/Core/DragHelper/index.js:
```diff
@@ -0,0 +1,44 @@
+
+  publicAPI.setLastOrigin = (origin) => {
+    // allow setting null as the last origin
+    if (origin && origin.length !== 3) return false;
```
shouldn't it trigger a vtkErrorMacro ?

finetjul commented on Sources/Widgets/Core/DragHelper/index.js:
```diff
@@ -0,0 +1,44 @@
+}
+
+const defaultValues = (initialValues) => ({
+  lastOrigin: null,
```
maybe rename into lastHandleOrigin or lastWidgetOrigin ?

Personal preference: I prefer "currentOrigin" vs "lastOrigin". For some reason working with a "last" information carries a notion of "obsolete" information.

finetjul reviewed (commented):
Some ideas:
 - find a way to move the dragHelper into the "manipulator" in order to 1) simplify the overall usage for "widgets developers" 2) group all "coordinate manipulation" into one class 3) simplify custom "drag" behaviors by changing/configuring the manipulator.
 - separate the "lastOrigin" from the "firstOriginClicked". It could be useful for some manipulators to know use that firstOriginClicked position.
 - I think our naming convention is a bit confusing. There are 3 words used interchangeably: "position", "coords" and "origin"

floryst commented:
I've decided to try something different in the 2nd commit: the manipulator now exposes a `worldDelta` property, which is the 3D vector difference between the previously handled event and the currently handled event. So far I've only done the [LineWidget](https://github.com/Kitware/vtk-js/pull/3073/files#diff-7fb5f9ba6a5b3f56d275bd25d13d359dd54322db636cd9beabfbd421a71b251a), [AngleWidget](https://github.com/Kitware/vtk-js/pull/3073/files#diff-841c9bbf0787ef517ac395498609b9961732c1855204fb980e3d8f205cef3b03), and [ImplicitPlaneWidget](https://github.com/Kitware/vtk-js/pull/3073/files#diff-77b18c2899d081bb11d53ee63049bbb8865a39f7731be80289f685813800de8a).

Generally speaking, the worldDelta should be sufficient for most dragging functionality. In constrained scenarios (e.g. see the ImplicitPlaneWidget code), an extra state variable is needed per constrained state.

I think this is sufficient for now. I'm not yet sure if we need to have the manipulators understand the startDrag/onDrag/endDrag lifecycle just yet. If you think this is sufficient, I'll change all of the widgets to use worldDelta.

floryst commented:
The second commit drops the drag helper in favor of exposing a `worldDelta` property on the manipulator's handleEvent() return value.

This doesn't work well with complex-constrained manipulations, e.g. the SeedWidget + PickerManipulator. In those scenarios, it's not the user manipulation that needs to be constrained, but rather the seed actor's position.

floryst commented:
@finetjul please take a look again!

finetjul commented on Sources/Widgets/Manipulators/AbstractManipulator/index.d.ts:
```diff
@@ -49,11 +49,20 @@ export interface vtkAbstractManipulator extends vtkObject {
      * 
+     * Returns an object containing:
+     * - worldCoords: a 3D coordinate corresponding to the 2D event. Can be null.
+     * - worldDelta: a 3D position delta between the current and the previous call to handleEvent.
```
please explain what is returned the first time it is called.

finetjul replied:
please explain that sometimes it is important to call handleEvent "just" to have worldDelta in correct state for next call.

finetjul commented on Sources/Widgets/Manipulators/AbstractManipulator/index.js:
```diff
@@ -27,6 +30,27 @@ function vtkAbstractManipulator(publicAPI, model) {
+
+  model._computeDeltaFromPrevCoords = (curWorldCoords) => {
+    if (!model._prevWorldCoords?.length || !curWorldCoords?.length)
+      return [0, 0, 0];
```
if _prevWorldCoords is empty/null, shouldn't it return curWorldCoords?
if curWorldCoords is empty/null, shouldn't it return _prevWorldCoords ?

floryst replied:
I think that if there's no previous world coordinate (e.g. first call), then returning [0,0,0] makes sense, since there's no movement. I can also see the case for returning `curWorldCoords` only in the case where `curWorldCoords` is not null, since it's interpreted as a delta from the origin, but that has a built-in assumption on the first call.

If `curWorldCoords` is null, then we shouldn't return the previous world coordinate. That would invalidate the concept of a delta entirely.

finetjul commented on Sources/Widgets/Widgets3D/AngleWidget/behavior.js:
```diff
@@ -85,20 +87,24 @@ export default function widgetBehavior(publicAPI, model) {
+        }
+        if (model._isDragging) {
+          model.activeState.setOrigin(
+            add(model.activeState.getOrigin(), worldDelta, [])
```
please explain in comments why here you use `worldDelta`, but not if `activeState === model.widgetState.getModeHandle)`. It is not obvious to me.

floryst replied:
The move handle can have a null origin, so `add(moveHandle.getOrigin(), worldDelta, [])` crashes. I'll add a comment about it.

finetjul commented on Sources/Widgets/Widgets3D/PolyLineWidget/behavior.js:
```diff
@@ -37,7 +38,15 @@ export default function widgetBehavior(publicAPI, model) {
-      model.activeState.setOrigin(worldCoords);
+      if (model._isDragging) {
+        // only use the drag helper if we are dragging a handle,
+        // not moving the move handle
```
I think this documentation could go in `handleEvent`, to know when to rely on worldDelta, and when not to rely on worldDelta. 
It is not obvious why dragging a handle requires a drag helper (which is an obsolete name ?)

floryst replied:
This is an obsolete comment that needs to be removed.

finetjul commented on Sources/Widgets/Widgets3D/ShapeWidget/behavior.js:
```diff
@@ -498,11 +498,19 @@ export default function widgetBehavior(publicAPI, model) {
+        model.point1 = vtkMath.add(
+          model.point1Handle.getOrigin(),
+          worldDelta,
+          []
```
`vtkMath.add(model.point1Handle.getOrigin(), worldDelta, model.point1)` to avoid array instanciation...

floryst commented:
I've pushed a new commit (which will be squashed prior to merging):
- updated worldDelta docs
- simplified worldDelta handling. The logic should now be more explicit about when we call `setOrigin(worldCoords)` (when the current origin is null) versus when we call `setOrigin(curOrigin + worldDelta)` (current origin is not null).

finetjul reviewed (approved):
LGTM
